### ISSUE TITLE
Incorrect url helpers being generated

### ISSUE THREAD

**Issue by gauravmanchanda** [phoenixframework/phoenix#2358]
### Environment

* Elixir version (elixir -v): 1.4.2
* Phoenix version (mix deps): 1.3.0-rc.2
* NodeJS version (node -v): 6.10.3
* NPM version (npm -v): 3.10.10
* Operating system: MacOS El Capitan 10.11.6

### Expected behavior

When I add a scoped route like this:

```
scope "/foo", Tattlr.Web.Foo do
  pipe_through :browser
  resources "/bar", BarController
end
```
The following URL Helpers should be generated

```
foo_bar_path  GET     /foo/bar             App.Web.Foo.BarController :index
foo_bar_path  GET     /foo/bar/:id/edit    App.Web.Foo.BarController :edit
foo_bar_path  GET     /foo/bar/new         App.Web.Foo.BarController :new
foo_bar_path  GET     /foo/bar/:id         App.Web.Foo.BarController :show
foo_bar_path  POST    /foo/bar             App.Web.Foo.BarController :create
foo_bar_path  PATCH   /foo/bar/:id         App.Web.Foo.BarController :update
```

### Actual behavior
This is being generated:

```
bar_path  GET     /foo/bar             App.Web.Foo.BarController :index
bar_path  GET     /foo/bar/:id/edit    App.Web.Foo.BarController :edit
bar_path  GET     /foo/bar/new         App.Web.Foo.BarController :new
bar_path  GET     /foo/bar/:id         App.Web.Foo.BarController :show
bar_path  POST    /foo/bar             App.Web.Foo.BarController :create
bar_path  PATCH   /foo/bar/:id         App.Web.Foo.BarController :update
```
The scope is totally being ignored.

**Comment by ryanwinchester** [comment]
what happens if you change:

```elixir
scope "/foo", Tattlr.Web.Foo do
```

to

```elixir
scope "/foo", Tattlr.Web.Foo, as: :foo do
```

**Comment by gauravmanchanda** [comment]
Hey Thanks @ryanwinchester, that works.

But I think there's an issue with the `phx.gen.html` task, coz in templates, the url helper used was `bar_path`, while in the controller, the url helper being used was `foo_bar_path`.

So, adding what you suggested did make the controller compile successfully, but template compilation failed due to unscoped paths being used.

**Comment by josevalim** [comment]
@gauravmanchanda if that's the case, that's an issue with the command. Which command did you run?

**Comment by gauravmanchanda** [comment]
@josevalim Yeah, maybe.
I ran this command

`mix phx.gen.html Foo Bar bar summary:string title:string  --binary-id --web Foo`

After it ran, it told me to add the following in the router:

```
scope "/foo", Tattlr.Web.Foo do
  pipe_through :browser
  ...
  resources "/bar", BarController
end
```

So, I see 2 issues here:
1) The route to be added doesn't tell me to add `as: :foo` in the scope definition.
2) The route helpers generated don't have the scope prepended to them, but the controller still uses routes with the scope prepended to them. In this case, the route helper generated was `bar_path`, but in the controller, `foo_bar_path` was being used.

**Comment by josevalim** [comment]
Ok, I guess we either need to document that if `--web` is used, you will need to change the paths, or generate something like `resources "/bar", BarController, as: :foo_bar`.

**Comment by gauravmanchanda** [comment]
Okay. Yeah, that sounds good.

We should make one change though. The url helpers being used in controllers/templates should be uniform. Currently there's a mismatch between them when using the `--web` flag.

What do you think?

**Comment by josevalim** [comment]
> We should make one change though. The url helpers being used in controllers/templates should be uniform. Currently there's a mismatch between them when using the --web flag.

What do you mean? Some of them are right and some are not?

**Comment by gauravmanchanda** [comment]
Yup. For instance, in the controller, the url helper used is `foo_bar_path` 

```elixir
def create(conn, %{"bar" => bar_params}) do
  case Foo.create_bar(bar_params) do
    {:ok, bar} ->
      conn
      |> put_flash(:info, "Bar created successfully.")
      |> redirect(to: foo_bar_path(conn, :show, bar))
    {:error, %Ecto.Changeset{} = changeset} ->
      render(conn, "new.html", changeset: changeset)
  end
end
```

while in the templates, its `bar_path`

```eex
<span><%= link "Edit", to: bar_path(@conn, :edit, @bar) %></span>
<span><%= link "Back", to: bar_path(@conn, :index) %></span>
```

**Comment by britto** [comment]
Some of these issues have been addressed at #2302. @gauravmanchanda have you tried running from master?

**Comment by gauravmanchanda** [comment]
Hey @britto, haven't tried running from `master`, working with `1.3.0-rc.2`. Just looked at the files changed in the commit and seems like the issues have been addressed. Thanks for the pointer :)

Would keep in mind to check things out on master the next time. Sorry for the duplicacy!!
